### PR TITLE
Fix #21356: JSON nested_to_record Silently Drops Top-Level None Values

### LINKED ISSUE #21356: JSON nested_to_record Silently Drops Top-Level None Values
xref https://github.com/pandas-dev/pandas/pull/21164#issuecomment-394510095

`nested_to_record` is silently dropping `None` values that appear at the top of the JSON. This is IMO unexpected and undesirable.

#### Code Sample, a copy-pastable example if possible

```python
In [3]: data = {
   ...:     "id": None,
   ...:     "location": {
   ...:         "country": None
   ...:     }
   ...: }

In [5]: nested_to_record(data)
Out[5]: {'location.country': None}
```
#### Problem description

The top level `None` value should not be dropped but rather preserved along with lower levels for consistency.

#### Expected Output
```python
In [5]: nested_to_record(data)
Out[5]: {'id': None, 'location.country': None}
```

Note this will break a few tests in `pandas/test_normalize.py`

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: ab6aaf73a848a8725a23bb880be5221dd5ef5b3d
python: 3.6.4.final.0
python-bits: 64
OS: Darwin
OS-release: 17.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0.dev0+67.gab6aaf73a
pytest: 3.4.1
pip: 10.0.1
setuptools: 38.5.1
Cython: 0.27.3
numpy: 1.14.1
scipy: 1.0.0
pyarrow: 0.8.0
xarray: 0.10.0
IPython: 6.2.1
sphinx: 1.7.0
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.1.2
openpyxl: 2.5.0
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.5
pymysql: 0.8.0
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: 0.1.3
fastparquet: 0.1.4
pandas_gbq: 0.4.1
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by daminisatya**
- [x] closes #21356 
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/json/test_normalize.py [hunk 1]
```diff
@@ -242,11 +242,13 @@ def test_missing_field(self, author_missing_data):
         # unnest records where only the first record has a None value
```
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: Can delete this comment (tbh I don't really understand what it means even with the current behavior)

File: pandas/tests/io/json/test_normalize.py [hunk 2]
```diff
@@ -351,7 +353,7 @@ def test_json_normalize_errors(self):
                       errors='raise'
                       )
 
-    def test_nonetype_dropping(self):
+    def test_nonetype(self):
         # GH20030: Checks that None values are dropped in nested_to_record
```
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: Can certainly delete this comment, though replace with a reference to the GitHub issue you are working on

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks really good for a first PR. Just some minor comments on the comments. Otherwise can you add a whatsnew note for 0.24?

File: pandas/tests/io/json/test_normalize.py [hunk 3]
```diff
@@ -238,15 +238,16 @@ def test_non_ascii_key(self):
         tm.assert_frame_equal(result, expected)
 
     def test_missing_field(self, author_missing_data):
-        # GH20030: Checks for robustness of json_normalize - should
-        # unnest records where only the first record has a None value
+        # GH20030: Checks for robustness of json_normalize
```
Comment (daminisatya) [inline pandas/tests/io/json/test_normalize.py]: @WillAyd Can you verify this comment once again.

File: pandas/tests/io/json/test_normalize.py [hunk 4]
```diff
@@ -351,9 +352,8 @@ def test_json_normalize_errors(self):
                       errors='raise'
                       )
 
-    def test_nonetype_dropping(self):
-        # GH20030: Checks that None values are dropped in nested_to_record
-        # to prevent additional columns of nans when passed to DataFrame
+    def test_nonetype(self):
+        # GH21356
```
Comment (daminisatya) [inline pandas/tests/io/json/test_normalize.py]: Here, i just added the issue number.

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -80,6 +80,8 @@ Documentation Changes
 Bug Fixes
 ~~~~~~~~~
 
+- The top level None value is not dropped but rather preserved along with lower levels for consistency (:issue:`21356`)
```
Comment (daminisatya) [inline doc/source/whatsnew/v0.24.0.txt]: Added the bugfix in the whats new note.

**Review by daminisatya — COMMENTED** [review]
Made the requested changes.

File: doc/source/whatsnew/v0.24.0.txt (hunk 5 above)
Comment (WillAyd) [reply doc/source/whatsnew/v0.24.0.txt]: Change of plans here - can you add to the "Fixed Regressions" section in v0.23.1 introduced via #21368?

File: pandas/tests/io/json/test_normalize.py (hunk 3 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_normalize.py]: Just the GH issue number is sufficient (like you have below)

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor edits otherwise lgtm

File: doc/source/whatsnew/v0.24.0.txt (hunk 5 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: this also needs to mention json_normalize, add double backticks on ``None``. 

finally, as a user this is not really clear on what has changed, can you re-word a bit.

File: pandas/tests/io/json/test_normalize.py [hunk 6]
```diff
@@ -351,9 +352,8 @@ def test_json_normalize_errors(self):
                       errors='raise'
                       )
 
-    def test_nonetype_dropping(self):
-        # GH20030: Checks that None values are dropped in nested_to_record
-        # to prevent additional columns of nans when passed to DataFrame
+    def test_nonetype(self):
```
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: can you make this test name more informative

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@WillAyd if you have a few, can you rebase this

**Comment by WillAyd** [comment]
@jreback think its good to go. First time doing that so let me know if anything looks off

**Comment by jorisvandenbossche** [comment]
@daminisatya Thanks a lot!

**Pull request merged**